documentation/general/dotnet-run-file.md

```diff
@@ -0,0 +1,224 @@
+# `dotnet run file.cs`
+
+This is a proposal for extending the dotnet CLI to allow running C# source files with no need for an explicit backing project.
+We call these *file-based programs* (as opposed to *project-based programs*).
+
+```ps1
+dotnet run file.cs
+```
+
+> [!NOTE]
+> This document describes the ideal final state, but the feature will be implemented in [stages](#stages).
+
+## Motivation
+
+File-based programs
+- should be a viable alternative to using PowerShell/bash scripts in .NET repos, and
+- lower the entry barrier for new customers.
+
+## Guiding principle
+
+The overarching guiding principle is that file-based programs have a simple and reliable [grow up](#grow-up) story to project-based programs.
+Previous file-based approaches like scripting are a variant of C# and as such have no simple and reliable grow up story.
+
+## Implicit project file
+
+The [guiding principle](#guiding-principle) implies that we can think of file-based programs as having an implicit project file.
+
+The implicit project file is the default project that would be created by running `dotnet new console`.
+This means that the behavior of `dotnet run file.cs` can change between SDK versions if the `dotnet new console` template changes.
+In the future we can consider supporting more SDKs like the Web SDK.
+
+## Grow up
+
+When file-based programs reach an inflection point where build customizations in a project file are needed,
+a single CLI command can be executed to generate a project file.
+In fact, this command simply materializes the [implicit project file](#implicit-project-file) on disk.
+This action should not change the behavior of the target program.
+
+```ps1
+dotnet project add
+```
+
+## Target path
+
+The path passed to `dotnet run ./some/path.cs` is called *the target path*.
+If it is a file, it is called *the target file*.
+*The target directory* is the directory of the target file, or the target path if it is not a file.
+
+We can consider adding an option like `dotnet run --from-stdin` which would read the C# file from the standard input.
+In this case, the current working directory would not be used to search for project or other C# files,
+the compilation would consist solely of the single file read from the standard input.
+
+## Integration into the existing `dotnet run` command
+
+`dotnet run file.cs` already has a meaning if there is a project file inside the current directory,
+specifically `file.cs` is passed as the first command-line argument to the target program.
+We preserve this behavior to avoid a breaking change.
+The file-based build and run kicks in only when:
+- a project file cannot be found (in the current directory or via the `--project` option), and
+- if the target path is a file, it has the `.cs` file extension, and
+- the target path (file or directory) exists.
+
+> [!NOTE]
+> This means that `dotnet run path` stops working when a file-based program [grows up](#grow-up) into a project-based program.
```

Comment:
Does dotnet run have a syntax for "the stuff after this point are not switches to dotnet run"? Something like how git and other Unix tools use "--" by itself to mean "everything after this point is not a switch". I assume there's something because otherwise how would you run your program and pass something that looks like a switch to it...

Reply:
Ah I see the answer is "yes" that line 75 mentions it, but it's unclear to me how it's interpreted.